### ISSUE TITLE
Error: not enough money (task MakerReserveOfferFunds)

### ISSUE THREAD

**Issue by fa2a5qj3** [haveno-dex/haveno#1089]
- Have available total of 12.33 XMR in haveno wallet.
- Create an offer to buy BTC, with 10 XMR.  Offer requires 11.55 XMR to be reserved.
- Select `Reserve only the funds needed.`
- Click through to complete the offer placement.

Error displayed is as follows:

```
Error
An error occurred when placing the offer:

not enough money (task MakerReserveOfferFunds)

No funds have left your wallet yet.
Please restart your application and check your network connection.

```

Log file:

```
Jul-02 16:43:59.692 [Thread-70] WARN  h.c.o.p.tasks.MakerReserveOfferFunds: Error creating reserve tx, attempt=2/5, offerId=VEQYKK, error=not enough money 
Jul-02 16:44:04.693 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: Creating reserve tx with preferred subaddress index=37, return address=86JLAk5SromHSZEnaYrBhoT65Uq8XDtt1PbbJSfs9V7fT6WLYbWXBsUJKNHC72memnaE2Pp2ZCSMfe7nBC556gd49UG5A7o 
Jul-02 16:44:04.702 [Thread-70] WARN  h.c.o.p.tasks.MakerReserveOfferFunds: Error creating reserve tx, attempt=3/5, offerId=VEQYKK, error=not enough money 
Jul-02 16:44:09.703 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: Creating reserve tx with preferred subaddress index=37, return address=86JLAk5SromHSZEnaYrBhoT65Uq8XDtt1PbbJSfs9V7fT6WLYbWXBsUJKNHC72memnaE2Pp2ZCSMfe7nBC556gd49UG5A7o 
Jul-02 16:44:09.712 [Thread-70] WARN  h.c.o.p.tasks.MakerReserveOfferFunds: Error creating reserve tx, attempt=4/5, offerId=VEQYKK, error=not enough money 
Jul-02 16:44:14.713 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: Creating reserve tx with preferred subaddress index=37, return address=86JLAk5SromHSZEnaYrBhoT65Uq8XDtt1PbbJSfs9V7fT6WLYbWXBsUJKNHC72memnaE2Pp2ZCSMfe7nBC556gd49UG5A7o 
Jul-02 16:44:14.717 [Thread-70] WARN  h.c.o.p.tasks.MakerReserveOfferFunds: Error creating reserve tx, attempt=5/5, offerId=VEQYKK, error=not enough money 
Jul-02 16:44:14.717 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: resetAddressEntriesForOpenOffer offerId=VEQYKK-cdc02eb3-0940-438e-8254-85d222c8d169-108 
Jul-02 16:44:14.718 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: swap addressEntry with address 87scEXpmDgt3c2kCS45Xi5eo2taBs3hSpFhjnG8enbkYHvkew6CUy47QY6QGBcbhpA9EYPK1kMFcAWAWLkyg5piu2BqifQY and offerId VEQYKK-cdc02eb3-0940-438e-8254-85d222c8d169-108 from context OFFER_FUNDING to available 
Jul-02 16:44:14.718 [Thread-70] INFO  haveno.core.xmr.wallet.XmrWalletService: swap addressEntry with address 86JLAk5SromHSZEnaYrBhoT65Uq8XDtt1PbbJSfs9V7fT6WLYbWXBsUJKNHC72memnaE2Pp2ZCSMfe7nBC556gd49UG5A7o and offerId VEQYKK-cdc02eb3-0940-438e-8254-85d222c8d169-108 from context TRADE_PAYOUT to available 
Jul-02 16:44:14.718 [Thread-70] ERROR haveno.common.taskrunner.Task: not enough money (task MakerReserveOfferFunds) monero.common.MoneroRpcError: not enough money
	at monero.common.MoneroRpcConnection.validateRpcResponse(MoneroRpcConnection.java:634)
	at monero.common.MoneroRpcConnection.sendJsonRequest(MoneroRpcConnection.java:393)
	at monero.common.MoneroRpcConnection.sendJsonRequest(MoneroRpcConnection.java:335)
	at monero.wallet.MoneroWalletRpc.createTxs(MoneroWalletRpc.java:1211)
	at monero.wallet.MoneroWalletDefault.createTx(MoneroWalletDefault.java:378)
	at monero.wallet.MoneroWalletRpc.createTx(MoneroWalletRpc.java:92)
	at haveno.core.xmr.wallet.XmrWalletService.createTx(XmrWalletService.java:458)
	at haveno.core.xmr.wallet.XmrWalletService.createTradeTxFromSubaddress(XmrWalletService.java:711)
	at haveno.core.xmr.wallet.XmrWalletService.createTradeTx(XmrWalletService.java:690)
	at haveno.core.xmr.wallet.XmrWalletService.createReserveTx(XmrWalletService.java:632)
	at haveno.core.offer.placeoffer.tasks.MakerReserveOfferFunds.run(MakerReserveOfferFunds.java:87)
	at haveno.common.taskrunner.TaskRunner.next(TaskRunner.java:68)
	at haveno.common.taskrunner.TaskRunner.handleComplete(TaskRunner.java:88)
	at haveno.common.taskrunner.Task.complete(Task.java:58)
	at haveno.core.offer.placeoffer.tasks.ValidateOffer.run(ValidateOffer.java:94)
	at haveno.common.taskrunner.TaskRunner.next(TaskRunner.java:68)
	at haveno.common.taskrunner.TaskRunner.run(TaskRunner.java:59)
	at haveno.core.offer.placeoffer.PlaceOfferProtocol.placeOffer(PlaceOfferProtocol.java:85)
	at haveno.core.offer.OpenOfferManager.signAndPostOffer(OpenOfferManager.java:1215)
	at haveno.core.offer.OpenOfferManager.lambda$doProcessUnpostedOffer$36(OpenOfferManager.java:944)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Error reproduced with varying numbers of UTXOs.

I am currently investigating a fix.

**Comment by woodser** [comment]
Are you connected to xmr-node.cakewallet.com by chance? Denoted by `monerod=` in the logs.

**Comment by woodser** [comment]
Other users have been seeing this, and I think it occasionally happens when connected to specific nodes over Tor.

Switching to clearnet fixed this user's error: https://github.com/haveno-dex/haveno/issues/1090#issuecomment-2204699076

**Comment by fa2a5qj3** [comment]
It is reproducible & happens with local or remote nodes.  What I did not realize before: it depends on the exact same amount having been received in a past tx, and is no longer an available UTXO (i.e. a prior offer for the same amount).  When tracing the code, I see it pick an **old** tx and use it here:

https://github.com/haveno-dex/haveno/blob/ec9f91e014a337c1baf21eeddf232f1f88672e09/core/src/main/java/haveno/core/offer/OpenOfferManager.java#L934

**New steps to reproduce from a fresh wallet:**

- Create an account for trading (e.g. Cyrptocurrencies, BTC)
- Create an offer to buy BTC with 10 XMR.
- Fund the offer from an external wallet with exactly 11.55 XMR as prompted.
- Cancel the offer, wait for the tx to unlock, then empty your Haveno wallet.

Now, back to the original description: fund your wallet as follows and let it confirm/unlock.

- Have available total of approx. 12.33 to 13.10 XMR in haveno wallet.
- Create an offer to buy BTC, with 10 XMR.  Offer requires 11.55 XMR to be reserved.
- Select `Reserve only the funds needed.`
- Click through to complete the offer placement.
- Error displayed.


While typing these revised steps, I verified the test case using a brand new wallet.
Complete wallet history for test case:

![image](https://github.com/haveno-dex/haveno/assets/174058787/e9126cad-c7de-4334-9242-497e4326c288)

**Comment by woodser** [comment]
Thanks for tracking down this difficult bug with reproducible steps. I was able to recreate and fix the error based on your steps: https://github.com/haveno-dex/haveno/pull/1091

This bit of code needed refactored anyway so this was the perfect chance.

**Comment by fa2a5qj3** [comment]
Resolved by #1091
